controllers/webhook.go

```diff
@@ -0,0 +1,166 @@
+		return err
+	}
+
+	if !hasK8sDSPAs {
+		log.Info("No other DSPAs with PipelineStorage 'kubernetes' found. Cleaning up webhook resources.")
+		if err := r.cleanupWebhookResources(ctx, params.DSPONamespace); err != nil {
+			log.Error(err, "Failed to clean up webhook resources")
+			return err
+		}
+	}
+
+	log.Info("Webhook resources cleanup complete.")
+	return nil
+}
+
+func (r *DSPAReconciler) cleanupWebhookResources(ctx context.Context, namespace string) error {
+	log := r.Log.WithValues("namespace", namespace)
+	webhookName := "pipelineversions.pipelines.kubeflow.org"
+	k8sWebhookName := "kubernetes-webhook"
+	webhookNamespaced := types.NamespacedName{Name: k8sWebhookName, Namespace: namespace}
+
+	// Delete MutatingWebhookConfiguration
+	mutating := &admv1.MutatingWebhookConfiguration{}
+	if err := r.Get(ctx, types.NamespacedName{Name: webhookName}, mutating); err == nil {
+		log.Info("Deleting MutatingWebhookConfiguration", "webhook", mutating)
+		if err := r.Delete(ctx, mutating); err != nil {
+			log.Error(err, "Failed to delete MutatingWebhookConfiguration")
+			return err
+		}
+	}
+
+	// Delete ValidatingWebhookConfiguration
+	validating := &admv1.ValidatingWebhookConfiguration{}
+	if err := r.Get(ctx, types.NamespacedName{Name: webhookName}, validating); err == nil {
+		log.Info("Deleting ValidatingWebhookConfiguration")
+		if err := r.Delete(ctx, validating); err != nil {
+			log.Error(err, "Failed to delete ValidatingWebhookConfiguration")
+			return err
+		}
+	}
+
+	// Delete Role
+	role := &rbacv1.ClusterRole{}
+	if err := r.Get(ctx, types.NamespacedName{Name: k8sWebhookName}, role); err == nil {
+		log.Info("Deleting webhook Role", "role", role)
+		if err := r.Delete(ctx, role); err != nil {
+			log.Error(err, "Failed to delete webhook Role")
+			return err
+		}
+	}
+
+	// Delete RoleBinding
+	roleBinding := &rbacv1.ClusterRoleBinding{}
+	if err := r.Get(ctx, types.NamespacedName{Name: k8sWebhookName}, roleBinding); err == nil {
+		log.Info("Deleting webhook RoleBinding", "roleBinding", roleBinding)
+		if err := r.Delete(ctx, roleBinding); err != nil {
+			log.Error(err, "Failed to delete webhook RoleBinding")
+			return err
+		}
+	}
+
+	// Delete Service
+	svc := &corev1.Service{}
+	if err := r.Get(ctx, webhookNamespaced, svc); err == nil {
+		log.Info("Deleting webhook Service")
+		if err := r.Delete(ctx, svc); err != nil {
+			log.Error(err, "Failed to delete webhook Service")
+			return err
+		}
+	}
+
+	// Delete ServiceAccount
+	sa := &corev1.ServiceAccount{}
+	if err := r.Get(ctx, webhookNamespaced, sa); err == nil {
+		log.Info("Deleting webhook ServiceAccount")
+		if err := r.Delete(ctx, sa); err != nil {
+			log.Error(err, "Failed to delete webhook ServiceAccount")
+			return err
+		}
+	}
+
+	// Delete deployments
+	var deployList appsv1.DeploymentList
+	if err := r.List(ctx, &deployList, &client.ListOptions{Namespace: namespace}); err != nil {
```

Comment:
Isn't there only ever one deployment? Using r.Get would be better here.

I'd also delete the deployment first since the deployment is still using the SA deleted just prior.